### ISSUE TITLE
single global queue?

### ISSUE THREAD

**Issue by cgwalters** [servo/homu#82]
I may be wrong about this, but I was investigating some slowness and am just now realizing that there's apparently a single *global* queue for builds across multiple repos.

```
def process_queue(...)
    for repo_label, repo in repos.items():
        for state in repo_states:
            if state.status == 'pending' and not state.try_:
                break
            elif state.status == '' and state.approved_by:
                if start_build_or_rebuild(state, repo_cfgs, buildbot_slots, logger, db, git_cfg):
                    return
```

Why do the `start_build()` cases use `return` in contrast to the first one which is just `break`?  This means if *any* build was started, we skip processing of any other ones.

**Comment by Manishearth** [comment]
I don't think that's the case? Servo's homu instance runs builds for multiple repos at once, at least.

queue_handler gets called from other places too. See https://github.com/servo/homu/blob/master/homu/server.py#L474

**Comment by cgwalters** [comment]
Yes, but AFAICS nothing requires that the build `queue_handler()` starts was the one that logically triggered the queue to be processed.

I think the problem may occur more obviously when an exception occurs when starting a build.  I've definitely noticed that before when on one repo, the github account wasn't authorized, we didn't even try any other builds.
